### PR TITLE
Use Hastus microservice to export routes

### PR DESCRIPTION
**Pull request by eleksi**
- Note: Hastus import/export microservice is not yet in docker-compose setup. It has to be started separately from its repository for this to work until in is added to docker-compose setup.

- - -
This change is [Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/441)

### CODE REVIEW

**Review by tk-vincit — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/441)* status: 0 of 8 files reviewed, 5 unresolved discussions (waiting on @eleksi)

___
*[`ui/src/api/hastus.ts` line 25 at r2](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLK_UU11aASVkMf75JO:-NLK_UU11aASVkMf75JP:b-embjgd) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/55fa2516cbf1371a25812b5af853cf2b7a80a74b/ui/src/api/hastus.ts#L25)):*
> ```XML
>   });
> 
> export const exportToHastus = async (
> ```

"exportRoutesToHastus" for clarity. (You can easily rename this when importing if you want to define "exportRoutesToHastus" name again: `import { exportRoutesToHastus:  exportToHastus } from '../../api/hastus';`
___
*[`ui/src/api/hastus.ts` line 28 at r2](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLKXsPv1zSYfNzHBcTs:-NLKXsPv1zSYfNzHBcTt:b-er6hpn) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/55fa2516cbf1371a25812b5af853cf2b7a80a74b/ui/src/api/hastus.ts#L28)):*
> ```XML
>   labels: string[],
>   priority: Priority,
>   observationDate: DateTime,
> ```

So, we have `ExportBody` interface defined, which is used for one purpose, and then we have this very similar inline definition, which has 2/3 same fields and is defined inline? 
This could be fixed by multiple ways, e.g.
- creating common interface and extending it where needed
- changing this api to take `observationDate` as ISO date (and also renaming it to `observationISODate` for clarity)



_Code quote:_
```XML
  labels: string[],
  priority: Priority,
  observationDate: DateTime,
```
___
*[`ui/src/components/routes-and-lines/search/ExportToolbar.tsx` line 61 at r2](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLKaNUl3auqv_k_5jsF:-NLKaNUl3auqv_k_5jsG:b-i4oix3) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/55fa2516cbf1371a25812b5af853cf2b7a80a74b/ui/src/components/routes-and-lines/search/ExportToolbar.tsx#L61)):*
> ```XML
>   const exportRoutes = () => {
>     exportRoutesToHastus(
>       isExportingAll ? searchResultRouteLabels : selectedRouteLabels,
> ```

This looks suspicious because neither `searchResultRouteLabels` nor `selectedRouteLabels` sounds like they include "all" routes. What is happening here?
___
*[`ui/src/hooks/routes/useExportRoutes.ts` line 15 at r2](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLKZs1y9eWukoGeVPem:-NLKZs1y9eWukoGeVPen:byrvd93) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/55fa2516cbf1371a25812b5af853cf2b7a80a74b/ui/src/hooks/routes/useExportRoutes.ts#L15)):*
> ```XML
>   const canExport = priorities?.length === 1;
> 
>   const exportRoutesToHastus = async (routeLabels: string[]) => {
> ```

You could make this less error-prone and easier to debug/maintain by this:
```
if (!canExport ) {
throw new Error(`Can't export routes, expected exactly 1 priority but got "${priorities}"`)
}
```
___
*[`ui/src/uiComponents/SimpleSmallButton.tsx` line 21 at r2](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLKad4c4bOGUmsPF57S:-NLKad4c4bOGUmsPF57T:b-qpazcm) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/55fa2516cbf1371a25812b5af853cf2b7a80a74b/ui/src/uiComponents/SimpleSmallButton.tsx#L21)):*
> ```XML
>     ? 'border-grey bg-white text-gray-900 hover:border-brand active:border-brand'
>     : 'border-brand bg-brand text-white hover:bg-opacity-50 active:bg-opacity-50';
>   const disabledClassNames = disabled ? 'pointer-events-none opacity-70' : '';
> ```

Do we really need this pointer-events-none className? Usually browsers should handle this kind of things with buttons automatically and if they won't do that, then we have accessibility error.

Btw, disabled button should most likely have `cursor: not allowed` cursor and NOT "pointer-events-none"

**Review by eleksi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/441)* status: 0 of 8 files reviewed, 5 unresolved discussions (waiting on @tk-vincit)

___
*[`ui/src/api/hastus.ts` line 25 at r2](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLK_UU11aASVkMf75JO:-NLKoyN11P46l4sCS0Cf:b-896fix) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/55fa2516cbf1371a25812b5af853cf2b7a80a74b/ui/src/api/hastus.ts#L25)):*
<details><summary><i>Previously, tk-vincit wrote…</i></summary><blockquote>

"exportRoutesToHastus" for clarity. (You can easily rename this when importing if you want to define "exportRoutesToHastus" name again: `import { exportRoutesToHastus:  exportToHastus } from '../../api/hastus';`
</blockquote></details>

Done.
___
*[`ui/src/api/hastus.ts` line 28 at r2](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLKXsPv1zSYfNzHBcTs:-NLKpdjb-ohdG55m9FFI:b-uum4p8) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/55fa2516cbf1371a25812b5af853cf2b7a80a74b/ui/src/api/hastus.ts#L28)):*
<details><summary><i>Previously, tk-vincit wrote…</i></summary><blockquote>

So, we have `ExportBody` interface defined, which is used for one purpose, and then we have this very similar inline definition, which has 2/3 same fields and is defined inline? 
This could be fixed by multiple ways, e.g.
- creating common interface and extending it where needed
- changing this api to take `observationDate` as ISO date (and also renaming it to `observationISODate` for clarity)

</blockquote></details>

Well the second one was not inline interface definition but just separate function params (not object). But I have changed it now that both the functions take an object as a parameter and created a common interface which is extended for both purposes. I cannot change the Hastus API so the ISO format Date is still called `observationDate`.
___
*[`ui/src/components/routes-and-lines/search/ExportToolbar.tsx` line 61 at r2](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLKaNUl3auqv_k_5jsF:-NLKppnO8MAUFG6fl6b3:bleb854) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/55fa2516cbf1371a25812b5af853cf2b7a80a74b/ui/src/components/routes-and-lines/search/ExportToolbar.tsx#L61)):*
<details><summary><i>Previously, tk-vincit wrote…</i></summary><blockquote>

This looks suspicious because neither `searchResultRouteLabels` nor `selectedRouteLabels` sounds like they include "all" routes. What is happening here?
</blockquote></details>

Well by "all" we mean all search result routes, without having to select any. I renamed the variable and added a comment for clarity.
___
*[`ui/src/hooks/routes/useExportRoutes.ts` line 15 at r2](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLKZs1y9eWukoGeVPem:-NLKqKPY11SeGioQG94x:b498c3v) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/55fa2516cbf1371a25812b5af853cf2b7a80a74b/ui/src/hooks/routes/useExportRoutes.ts#L15)):*
<details><summary><i>Previously, tk-vincit wrote…</i></summary><blockquote>

You could make this less error-prone and easier to debug/maintain by this:
```
if (!canExport ) {
throw new Error(`Can't export routes, expected exactly 1 priority but got "${priorities}"`)
}
```
</blockquote></details>

This function should not be called if there are multiple priorities selected, as the button is disabled in that case. But maybe there will be bug somewhere or this function is called in some other context, so this error could be thrown. So I added it.
___
*[`ui/src/uiComponents/SimpleSmallButton.tsx` line 21 at r2](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLKad4c4bOGUmsPF57S:-NLKrKVc6PPXsV8-Np_f:bpj1q41) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/55fa2516cbf1371a25812b5af853cf2b7a80a74b/ui/src/uiComponents/SimpleSmallButton.tsx#L21)):*
<details><summary><i>Previously, tk-vincit wrote…</i></summary><blockquote>

Do we really need this pointer-events-none className? Usually browsers should handle this kind of things with buttons automatically and if they won't do that, then we have accessibility error.

Btw, disabled button should most likely have `cursor: not allowed` cursor and NOT "pointer-events-none"
</blockquote></details>

This is just implemented the same way SimpleButton is. The button is disabled even without the styles, but we also want to make it look disabled.

The reason we have `pointer-events-none` here is because we want to disable e.g. hover effects for this button when it is disabled. If we pass only `cursor: not allowed`, we would still see button corners turn blue when hovering over it. Is we pass both classes, `cursor: not allowed` does nothing as pointer effects are disabled.

**Review by tk-vincit — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/441)* status: 0 of 8 files reviewed, 2 unresolved discussions (waiting on @eleksi)

___
*[`ui/src/api/hastus.ts` line 28 at r2](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLKXsPv1zSYfNzHBcTs:-NLL9g-u4N2rsdbvoxbi:bpwasox) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/55fa2516cbf1371a25812b5af853cf2b7a80a74b/ui/src/api/hastus.ts#L28)):*
<details><summary><i>Previously, eleksi wrote…</i></summary><blockquote>

Well the second one was not inline interface definition but just separate function params (not object). But I have changed it now that both the functions take an object as a parameter and created a common interface which is extended for both purposes. I cannot change the Hastus API so the ISO format Date is still called `observationDate`.
</blockquote></details>

What do you mean "cannot change hastus api"? You don't have to, you can use whatever name in our own code and then rename it just in time before sending it to api.
```
{
 observationDate: observationISODate.
}
```

But OK, as long as `ExportParams` is used just in this file, there is not much room for confusion (at least so far) so this is good enough for now.
___
*[`ui/src/uiComponents/SimpleSmallButton.tsx` line 21 at r2](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLKad4c4bOGUmsPF57S:-NLLAMYQ6WOx7i4xPDHK:bqszqzq) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/55fa2516cbf1371a25812b5af853cf2b7a80a74b/ui/src/uiComponents/SimpleSmallButton.tsx#L21)):*
<details><summary><i>Previously, eleksi wrote…</i></summary><blockquote>

This is just implemented the same way SimpleButton is. The button is disabled even without the styles, but we also want to make it look disabled.

The reason we have `pointer-events-none` here is because we want to disable e.g. hover effects for this button when it is disabled. If we pass only `cursor: not allowed`, we would still see button corners turn blue when hovering over it. Is we pass both classes, `cursor: not allowed` does nothing as pointer effects are disabled.
</blockquote></details>

Tried to google this a bit, and seems like `cursor: not-allowed !important` might do the trick: https://stackoverflow.com/questions/50349017/how-can-i-change-cursor-for-disabled-button-or-a-in-bootstrap-4
Anyway, proper solution could actually be then extracting hover styles to own function:

```
const getHoverStyles = (disabled, inverted) => {
if (disabled) {
 return ''
}
return inverted ? ...;

const colorClassNames = `... ${getHoverStyles(disabled, inverted)}`;
```

**Review by eleksi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/441)* status: 0 of 11 files reviewed, 2 unresolved discussions (waiting on @tk-vincit)

___
*[`ui/src/api/hastus.ts` line 28 at r2](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLKXsPv1zSYfNzHBcTs:-NLLFZNFCBfrHpkOG37Q:b-e4s2lb) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/55fa2516cbf1371a25812b5af853cf2b7a80a74b/ui/src/api/hastus.ts#L28)):*
<details><summary><i>Previously, tk-vincit wrote…</i></summary><blockquote>

What do you mean "cannot change hastus api"? You don't have to, you can use whatever name in our own code and then rename it just in time before sending it to api.
```
{
 observationDate: observationISODate.
}
```

But OK, as long as `ExportParams` is used just in this file, there is not much room for confusion (at least so far) so this is good enough for now.
</blockquote></details>

I know we can use whatever name we want, but `ExportBody` is representing the Hastus API endpoint body schema, and `exportRoutes` was accepting the request body, aka. `ExportBody` as only parameter. So if we want to use `observationISODate` field, we cannot use it in this particular interface.

Of course we could change the function responsibilities so that exportRoutes takes something else than the request body as parameter and then renames the field from `observationISODate` to `observationDate`, but renaming / type converting is what `exportRoutesToHastus` is responsible for right now. I think it is nice to keep the actual HTTP call function as simple as possible so it does not have to rename object fields / convert types etc. before calling, even though in this case it probably wouldn't be too bad idea to combine `exportRoutes` and `exportRoutesToHastus` into one function either.

Anyway, good that this will suffice for now.
___
*[`ui/src/uiComponents/SimpleSmallButton.tsx` line 21 at r2](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLKad4c4bOGUmsPF57S:-NLLUR1S9VpDcH2eHpFb:b2xa34l) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/55fa2516cbf1371a25812b5af853cf2b7a80a74b/ui/src/uiComponents/SimpleSmallButton.tsx#L21)):*
<details><summary><i>Previously, tk-vincit wrote…</i></summary><blockquote>

Tried to google this a bit, and seems like `cursor: not-allowed !important` might do the trick: https://stackoverflow.com/questions/50349017/how-can-i-change-cursor-for-disabled-button-or-a-in-bootstrap-4
Anyway, proper solution could actually be then extracting hover styles to own function:

```
const getHoverStyles = (disabled, inverted) => {
if (disabled) {
 return ''
}
return inverted ? ...;

const colorClassNames = `... ${getHoverStyles(disabled, inverted)}`;
```
</blockquote></details>

Yes. I noticed also that there is quite a lot of code duplication and string concatenation in `SimpleButton` and `SimpleSmallButton` when it comes to classNames, and it has come quite hard to follow how `classNames` is constructed and which style is applied and when, especially in `SimpleButton`. 

So, I tried to refactor button classname generation a little bit and make it reusable. Please note that adding separate hover styles function as you suggested above would have made the original code even more lengthy, which is not visible straight away from the diffs.

I really can't decide whether or not this change takes it to a good or bad direction, there are pros and cons. What do you think?

**Review by tk-vincit — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/hsldevcom/jore4-ui/441)* status: 0 of 11 files reviewed, 1 unresolved discussion (waiting on @eleksi)

___
*[`ui/src/uiComponents/SimpleButton.tsx` line 28 at r5](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLQ6HBoCXtNt5tfdVEe:-NLQ6HBoCXtNt5tfdVEf:b7a90py) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/b13a33ec4937d9d474d3bbca1af714a8e6ff42dd/ui/src/uiComponents/SimpleButton.tsx#L28)):*
> ```XML
> 
> export const commonHoverStyle = 'hover:border-2 hover:border-tweaked-brand';
> const hoverStyle = `${commonHoverStyle} m-px hover:m-0`;
> ```

Could this be defined inside `getHoverStyles`? Its a bit confusing that we have both "hoverStyle" constant and also "getHoverStyles" function.
___
*[`ui/src/uiComponents/SimpleSmallButton.tsx` line 21 at r2](https://reviewable.io/reviews/hsldevcom/jore4-ui/441#-NLKad4c4bOGUmsPF57S:-NLQ6Win0VCs7lGRQ4iZ:b-2jdvxz) ([raw file](https://github.com/hsldevcom/jore4-ui/blob/55fa2516cbf1371a25812b5af853cf2b7a80a74b/ui/src/uiComponents/SimpleSmallButton.tsx#L21)):*
<details><summary><i>Previously, eleksi wrote…</i></summary><blockquote>

Yes. I noticed also that there is quite a lot of code duplication and string concatenation in `SimpleButton` and `SimpleSmallButton` when it comes to classNames, and it has come quite hard to follow how `classNames` is constructed and which style is applied and when, especially in `SimpleButton`. 

So, I tried to refactor button classname generation a little bit and make it reusable. Please note that adding separate hover styles function as you suggested above would have made the original code even more lengthy, which is not visible straight away from the diffs.

I really can't decide whether or not this change takes it to a good or bad direction, there are pros and cons. What do you think?
</blockquote></details>

After couple of iteration rounds this starts to look OK. :+1:

**Comment by eleksi** [comment]
e2e-tests that fail in CI work when run locally. They fail in CI apparently because of map-matching -response times out, which does not happen when test are run locally. Therefore, this PR doesn't seem to break any tests, the CI pipe is just broken, thus this PR can be merged.
